pkg/catalog/helm/setup.go

```diff
@@ -0,0 +1,253 @@
+package helm
+
+import (
+	"crypto/md5"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/rancher/rancher/pkg/catalog/git"
+	nsutil "github.com/rancher/rancher/pkg/namespace"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/sirupsen/logrus"
+)
+
+func New(namespace string, catalog *v3.Catalog) (*Helm, error) {
+	h, err := NewNoUpdate(namespace, catalog)
+	if err != nil {
+		return nil, err
+	}
+	if err := h.setup(); err != nil {
+		h.Unlock()
+		return nil, err
+	}
+	return h, nil
+}
+
+func NewNoUpdate(namespace string, catalog *v3.Catalog) (*Helm, error) {
+	if catalog == nil || catalog.Name == "" {
+		return nil, errors.New("Catalog is not defined for helm")
+	}
+
+	h := newCache(namespace, catalog.Name)
+	h.url = catalog.Spec.URL
+	h.branch = catalog.Spec.Branch
+	h.lastCommit = catalog.Status.Commit
+	h.username = catalog.Spec.Username
+	h.password = catalog.Spec.Password
+	h.Lock()
```

Comment:
Is it better to move the lock into specific read and write of cache files? i feel a little bit lost on where the lock is acquired and released...

Reply:
I agree the locking stuff can use some rework. However, whenever we create a New helm object we can assume there will be some I/O involved and we could Lock by default and Unlock when done. When using scaledContext to destroy, the lock does not seem to be effective with traverseAndUpdate (when modified appropriately), so may want to move to file locking instead.

Reply:
Actually, traverseAndUpdate is probably a special case where we need to check if the cache has been destroyed after re-locking, since the process is long-lived.

Reply:
Locking stuff has been revamped